_posts/2025-01-08-working-with-graph-databases-in-python.md

```diff
@@ -0,0 +1,278 @@
+comes from graph theory in mathematics, where a network of nodes representing data are connected using edges.
+This makes a graph database more flexible than a typical hierarchical database,
+as new nodes can be added and connected to any existing node, often without constraints.
+This lack of constraints can be convenient, but it does mean that a user will need
+additional context regarding the existing nodes before adding new ones, as the existing
+data is unpredictable.
+Constraints can be added to a graph database using additional tools if required.
+For example, when using
+a [Resource Description Framework (RDF)](https://en.wikipedia.org/wiki/Resource_Description_Framework)
+graph, you can add constraints
+using [Shapes Constraint Language (SHACL)](https://en.wikipedia.org/wiki/SHACL).
+
+In a graph database, everything is represented using URIs (Uniform Resource Identifier) or literals.
+The URIs are what allows us to create new connections between nodes in the graph,
+and even connections between multiple graphs.
+For this to work, data is stored in `triples`. In a triple, we have 3 pieces of information:
+- Subject: the row in a typical database, referring to one feature (URI)
+- Predicate: the column in a typical database, referring to one attribute name (URI)
+- Object: the value found at the given row/column, the attribute value (URI or literal)
+
+The predicate URI should also be a subject itself, containing more triples that describe its definition.
+This means that all of the context behind the data is described within the graph database,
+an advantage which cannot be done in a relational database.
+
+Let's apply this concept to a lithology example. If we want to represent Rhyolite's lithology classification
+as an Igneous Rock in a triple, we could do the following:
+- Subject: URI to Rhyolite
+- Predicate: URI to Parent Definition
+- Object: URI to Igneous Rock
+
+The value from this data structure comes from the fact that in this example, we could find the predicates
+of Igneous Rock using its own URI, and then continue to another subject after that.
+
+The nature of `triples` also means that when you iterate over the `triples` in a
+graph database, you get a single predicate object for a single subject at a time.
+You do not get all of the predicates for a subject at once.
+Therefore, you will likely want to parse the data into a different structure
+to suit your requirements, depending on what you wish to do with the data.
+
+It is important to note that because predicates are defined using a URI,
+we can use pre-defined predicates from online `namespaces`.
+In the example below, we will see many attributes from the online `skos` namespace,
+such as `skos:prefLabel` which is essentially a human readable name,
+and `skos:broader` which is a parent node.
+
+You can find more information about the `skos` data model [here](https://www.w3.org/2009/08/skos-reference/skos.html).
+
+## Why are we working with graph databases?
```

Comment:
We are using graph databases elsewhere too, so perhaps change this to "Use of graph data for lithology classifications in field data capture system"